src/PlottingInterface.cc

```diff
@@ -0,0 +1,542 @@
+}
+
+//////////////////////////////////////////////////////
+float PlottingInterface::Timeout() const
+{
+  return this->dataPtr->timer->interval();
+}
+
+//////////////////////////////////////////////////////
+float PlottingInterface::Time() const
+{
+  return this->dataPtr->time;
+}
+
+//////////////////////////////////////////////////////
+void PlottingInterface::onComponentSubscribe(QString _entity, QString _typeId,
+                                             QString _type, QString _attribute,
+                                             int _chart)
+{
+  // convert the strings into
+  uint64_t entity, typeId;
+  std::istringstream issEntity(_entity.toStdString());
+  issEntity >> entity;
+  std::istringstream issTypeId(_typeId.toStdString());
+  issTypeId >> typeId;
+
+  emit this->ComponentSubscribe(entity, typeId, _type.toStdString(),
+                                _attribute.toStdString(), _chart);
+}
+
+//////////////////////////////////////////////////////
+void PlottingInterface::onComponentUnSubscribe(QString _entity, QString _typeId,
+                                               QString _attribute, int _chart)
+{
+  // convert the strings into
+  uint64_t entity, typeId;
+  std::istringstream issEntity(_entity.toStdString());
+  issEntity >> entity;
+  std::istringstream issTypeId(_typeId.toStdString());
+  issTypeId >> typeId;
+
+  emit this->ComponentUnSubscribe(entity, typeId,
+                                  _attribute.toStdString(), _chart);
+}
+
+//////////////////////////////////////////////////////
+void PlottingInterface::subscribe(int _chart,
+                                  QString _topic,
+                                  QString _fieldPath)
+{
+  this->dataPtr->transport->Subscribe(_topic.toStdString(),
+                                      _fieldPath.toStdString(),
+                                      _chart);
+}
+
+////////////////////////////////////////////
+void PlottingInterface::InitTimer()
+{
+  this->dataPtr->timer = new QTimer();
+  this->dataPtr->timer->setInterval(this->dataPtr->timeout);
+  connect(this->dataPtr->timer, SIGNAL(timeout()), this, SLOT(UpdateGui()));
+  this->dataPtr->timer->start();
+
+  auto moveTimer = new QTimer();
+  moveTimer->setInterval(1000);
+  connect(moveTimer, SIGNAL(timeout()), this, SLOT(moveCharts()));
+}
+
+////////////////////////////////////////////
+void PlottingInterface::UpdateGui()
+{
+  this->dataPtr->transport->UnsubscribeOutdatedTopics();
+
+  auto topics = this->dataPtr->transport->Topics();
+
+  // Complexity O(Num of Dragged Items) or O(Num of Chart Value Axes)
+  for (auto const &topic : topics)
+  {
+    auto fields = topic.second->Fields();
+
+    for (auto const &field : fields)
+    {
+      auto charts = field.second->Charts();
+
+      for (auto const &chart : charts)
+      {
+        QString fieldFullPath = QString::fromStdString(
+          topic.first + "-" + field.first);
+        double x = this->dataPtr->time;
```

Comment:
We already spoke about this, but I'll just mention here for completeness that we could use the time from the message's header if that's present.